geps/gep-1911.md

```diff
@@ -0,0 +1,260 @@
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## TLDR
+
+Not all implementations support automatic protocol selection. Even in some cases protocols are disabled without an explicit opt-in (eg. websockets with Contour & NGINX). Thus application developers need the ability to specify the protocol(s) that their application supports.
+
+## Goals
+
+- Support protocols that can have a Gateway `*Route` resource as a frontend
+- Standardize Gateway API implementations on the protocols & constants defined by the Kubernetes [Standard Application Protocols (KEP-3726)](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3726-standard-application-protocols#summary)
+- Support backends with multiple protocols on the same port (ie. h2c/http1)
+
+## Non-Goals
+
+- Backend TLS (covered in [GEP-1897](https://github.com/kubernetes-sigs/gateway-api/issues/1897))
+- Additional protocol specific configuration
+- Disabling Protocols
+
+## Introduction
+
+Since Kubernetes 1.20 the `core/v1.Service` and `core/v1.EndpointSlice` object has a stable `appProtocol` field. It's purpose is to allow end-users to specify an application protocol (L7) for each service port.
+
+Originally the use of this field in the Gateway API was rejected in [GEP-1282](geps/gep-1282.md#non-goals):
+> v1.Service’s appProtocol field is not fit for purpose, because it is defined as accepting values either from the IANA Service Name registry, or domain-prefixed values and we need more flexibility than that. 
+
+Since then a Kubernetes enhancement proposal was created [KEP-3726](https://github.com/kubernetes/enhancements/issues/3726) to repurpose `appProtocol` to include a convention for protocols that are not IANA service names. This would involve prefixing protocol names with `kubernetes.io/*`. 
+
+Note: Kubernetes will automatically create `EndpointSlices` for `Services` that have a selector. [Custom `EndpointSlices`](https://kubernetes.io/docs/concepts/services-networking/service/#custom-endpointslices) can manually be created.
+
+## API Semantics
+
+A Gateway implementation MUST recognize the Kubernetes Standard Application Protocols ([KEP-3726](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3726-standard-application-protocols)) for specifying the protocol for a backend reference in a Gateway API `*Route` resource
+
+Thus when a `*Route` points to a Kubernetes Service resource the backend protocol for each port can be specified by
+- setting the `appProtocol` field on the Kubernetes `Service`
+- setting the `appProtocol` field on an `Endpoint`/`EndpointSlice` object associated with a Kubernetes `Service`
+
+At the moment there exists two defined constants:
+- `kubernetes.io/h2c` - HTTP/2 over cleartext as described in https://www.rfc-editor.org/rfc/rfc7540
+- `kubernetes.io/raw` - Instructs the implementation to look at the `Service`'s `protocol` field which supports `TCP`, `UDP`, and `SCTP`. Default is `TCP`.
```

Comment:
`kubernetes.io/raw` - was not recognized  as a standard (yet). I know the current situation is confusing because https://github.com/kubernetes/enhancements/pull/3912 was not merge yet to reflect the actual implementation. 
I will ping again there to ask for approval.

Reply:
PR merged so going to resolve this

Reply:
can you remove `kubernetes.io/raw`? it is not a standard so it may confuse other people.

Reply:
What's the alternative mechanic to say we support TCP etc. ?

Reply:
The current situation is the if the protocol is `TCP` and `appProtocol` is not set, implementations do whatever they want, one can take is "raw" TCP other can try to sniff and guess the traffic.

We wanted to introduce an appProtocol value that will basically be "raw" not only for TCP but for all/most port protocols (TCP/UDP/SCTP) hence the name `raw` and not `tcp-raw` or `tcp`.  That said, it was last minute and we did not added it eventually. (https://github.com/kubernetes/kubernetes/pull/115433#discussion_r1136267401)

You could add it to your current open enhancement and/or start a discussion in slack/github to see what other people think.

Reply:
I added a PR to add support for raw - https://github.com/kubernetes/enhancements/pull/4106

Reply:
Resolving as raw is not supported upstream - https://github.com/kubernetes/enhancements/pull/4106

Will remove it from the GEP